### ISSUE TITLE
[out-of-proc] `RetryContext.LastFailure` incorrectly captures thrown exception

### ISSUE THREAD

**Issue by Ilia-Kosenkov** [Azure/azure-functions-durable-extension#2711]
### Description
When inspecting `RetryContext.LastFailure` during activity retry cycle, I noticed that `.ErrorType` and `.ErrorMessage` are populated from `Exception.Message`, with `.ErrorType` being the first part of `.Message`, until a `':'`, while the remaining part of the `.Message` is copied to `.ErrorMessage`.

As a result of this, `RetryContext.LastFailure.IsCausedBy<TException>()` does not work, because it relies on `.ErrorType`, which is malformed.

If I drop `':'` from my exception message, then `.ErrorType` is `"(unknown)"`

### Expected behavior

- `RetryContext.LastFailure.ErrorType` is set to correct the exception type.
- `RetryContext.LastFailure.ErrorMessage` is set to `.Message` of the captured exception.
- `RetryContext.LastFailure.IsCausedBy<TException>()` works

### Actual behavior
- `RetryContext.LastFailure.ErrorType` seems to be set to `exception.Message` sliced till first `':'` or `"(unknown)"` if `':'` is absent
-  `RetryContext.LastFailure.ErrorMessage` seems to be set to `exception.Message` sliced from first `':'` or the beginning of the message
-  `RetryContext.LastFailure.InnerFailure` is `null` (cannot judge if this is useful or not)
- `RetryContext.LastFailure.IsCausedBy<TException>()` does not work

### Relevant source code snippets
Had to scramble some info here. 
Here is the exception I throw from my activity:
<img width="406" alt="Screenshot 2024-01-08 141740" src="https://github.com/Azure/azure-functions-durable-extension/assets/8782986/9c7d953b-f9f5-49f4-84dd-45441bb94a15">

And here is what I observe in the `AsyncRetryHandler`:
<img width="297" alt="Screenshot 2024-01-08 141810" src="https://github.com/Azure/azure-functions-durable-extension/assets/8782986/5f7c493c-02f1-4493-bd70-e5bf65d57922">

### Known workarounds
I originally intended to append `Reason: ...` to exception messages as a workaround, but this lead me to discovery that `':'` is treated specially (which made things even worse).
I am considering (ab-)using this feature to simulate the intended behavior.

### App Details
- `Microsoft.Azure.Functions.Worker.Extensions.DurableTask 1.1.0`
- `Microsoft.Azure.Functions.Worker.Sdk 1.16.4`
- Azure Functions v4
- .NET 8

**Comment by Ilia-Kosenkov** [comment]
I believe now it is a duplicate (or rather 'caused by') #2697. Please verify and close if it is so.

**Comment by davidmrdavid** [comment]
I agree the issues are related - we're looking to reconsider how we're treating inner exception handling in .NET isolated. Let me close this so we can use the issue you linked as the 'aggregate' ticket. Thanks

**Comment by cgillum** [comment]
Reactivating since I think this issue and #2697 are not actually related.

**Comment by j0hnth0m** [comment]
I continue to experience this issue  using all the latest nuget packages installed.  In my custom retry handler I receive a an 'ErrorType' that is derived from the original exception's message rather than the type name of the original exception thrown from the activity.

My handler is expecting to see "HttpRequestException" as the "ErrorType" instead I see "Response status code does not indicate success" as the "ErrorType" and "400" as the "ErrorMessage".  It appears that these properties are being set from the original exception's message split on the ":" character where the original exception message was: "Response status code does not indicate success: 400".

Is this a bug, if so, is there a fix or at least an open issue to address it that I can follow.  

Otherwise the only work-around I can see is to create a custom exception that takes advantage of this bug by setting the type in the exception message, followed by a ":", followed by the actual exception message as derived from the original exception.  Then the retry handler will get an ErrorType and ErrorMessage that are actually predictable.  Of course this would be extremely vulnerable to future updates should this issue finally get fixed.

Please advise.

**Comment by iansall** [comment]
@j0hnth0m I have the same issue with the latest NuGet packages - my workaround is to set EnableUserCodeException to false in HostBuilder.ConfigureFunctionsWorkerDefaults (even though this is now deprecated). As I understand it this was false by default until  recently when it switched to true by default.

**Comment by j0hnth0m** [comment]
> [@j0hnth0m](https://github.com/j0hnth0m) I have the same issue with the latest NuGet packages - my workaround is to set EnableUserCodeException to false in HostBuilder.ConfigureFunctionsWorkerDefaults (even though this is now deprecated). As I understand it this was false by default until recently when it switched to true by default.

Agree that doing this will preserve the original exception type as "ErrorType" however there are still other issues such as the original exception message getting truncated at the first occurance of a line ending char sequenece such as "\r\n",  any after this is lost. 

Also, any inner-exception information is completely lost.

This whole exception handling/marshalling between activities and orchestrations is a complete mess.  At this point I'm forced to trap all execptions in the activity and package them up and make it part of the function return value.  Feels like I'm back in the old 'C' HRESULT days.

**Comment by andystaples** [comment]
@j0hnth0m , @iansall I have an open PR for adding end-to-end tests for exception handling and retries in dotnet-isolated. [link](https://github.com/Azure/azure-functions-durable-extension/pull/3034/files) I'll add some tests covering these cases and see if I can't figure out what is going on.
